// @strapi/generators/dist/index.mjs
import { join } from "node:path";
import nodePlop from "node-plop";
const runCLI = async () => {
  const { Plop, run } = await import("plop");
  Plop.prepare(
    {
      configPath: join(__dirname, "plopfile.js")
    },
    (env) => {
      const argv = process.argv.slice(2);
      Plop.execute(env, argv, (env2, argv2) => {
        const options = {
          ...env2,
          dest: join(process.cwd(), "src")
          // this will make the destination path to be based on the cwd when calling the wrapper
        };
        return run(options, argv2, true);
      });
    }
  );
};
const generate = async (generatorName, options, { dir = process.cwd(), plopFile = "plopfile.js" } = {}) => {
  const plop = nodePlop(join(__dirname, plopFile), {
    destBasePath: join(dir, "src"),
    force: false
  });
  const generator = plop.getGenerator(generatorName);
  await generator.runActions(options, {
    onSuccess() {
    },
    onFailure() {
    },
    onComment() {
    }
  });
};
export {
  generate,
  runCLI
};
